.github/workflows/gradle.yml

```diff
@@ -15,21 +15,43 @@ jobs:
       - name: Check images exist
         run: ./tools/bin/check_images_exist.sh
 
-      - name: Cache java deps
+      - name: Docker Caching
         uses: actions/cache@v2
         with:
-          path: ~/.gradle
-          key: gradle-${{ hashFiles('**/*.gradle') }}
+          path: |
+            /tmp/docker-registry
+          key: docker-${{ runner.os }}-${{ hashFiles('Dockerfile') }}-${{ github.sha }}
           restore-keys: |
-            gradle-
+            docker-${{ runner.os }}-${{ hashFiles('Dockerfile') }}-
+            docker-${{ runner.os }}-
 
-      - name: Cache node deps
+      - name: Pip Caching
         uses: actions/cache@v2
         with:
-          path: ~/.npm
-          key: node-${{ hashFiles('**/package-lock.json') }}
+          path: |
+            ~/.cache/pip
+          key: pip-${{ runner.os }}-${{ hashFiles('**/setup.py') }}-${{ hashFiles('**/requirements.txt') }}
           restore-keys: |
-            node-
+            pip-${{ runner.os }}-
+
+      - name: Npm Caching
+        uses: actions/cache@v2
+        with:
+          path: |
+            ~/.npm
+            **/node_modules
+          key: npm-${{ runner.os }}-${{ hashFiles('**/package-lock.json') }}
+          restore-keys: |
+            npm-${{ runner.os }}-
+
+      - name: Gradle and Python Caching
+        uses: actions/cache@v2
+        with:
+          path: |
+            ~/.gradle/caches
```

Comment:
The Github Action cache examples prefer to cache the specific directories for gradle instead of the entire `~/.gradle`. 

This intentionally does not use restore-key.

Reply:
add comment saying this "This intentionally does not use restore-key."?